Grafana/UI: Add disable prop to Segment

oscarkilhed opened the pull request:
SegmentSync, SegmentAsync and SegmentInput had the disable prop inherited from HTMLProp but it did not disable the component. The disable prop should now disable the component.

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**Which issue(s) this PR fixes**:

Fixes #29573

oscarkilhed committed:
Grafana/UI: Add disable prop to Segment

SegmentSync, SegmentAsync and SegmentInput had the disable prop inherited from HTMLProp but it did not disable the component. The disable prop should now disable the component.

oscarkilhed committed:
Prettify with prettier

torkelo commented on packages/grafana-ui/src/components/Segment/SegmentInput.tsx:
```diff
@@ -34,9 +36,18 @@ export function SegmentInput<T>({
           Component || (
-            <a className={cx('gf-form-label', 'query-part', !value && placeholder && 'query-placeholder', className)}>
+            <a
+              className={cx(
+                'gf-form-label',
+                'query-part',
+                !value && placeholder && 'query-placeholder',
+                getStyles<T>({ disabled: disabled }).link,
+                className
```
fees like it would be better to use conditional classnames here: https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Button/ToolbarButton.tsx#L42 
Then you can move the getStyles call to the top (dont pass any props), just return a style with a "disabled"   style on it. then compose all the classses using cx like

```typescript
cx(
 'gf-form-label',
  'query-part', {
     ['query-placeholder']: !value && placeholder,
     [styles.disabled]: disabled
  },
className);
```

oscarkilhed replied:
Thanks, I wasn't quite happy with how the styles were applied. This looks cleaner, I'll get on it.

dprokop replied:
Good point @torkelo

oscarkilhed committed:
Simplify how styles/classnames are applied to segment components

Clarity-89 commented on packages/grafana-ui/src/components/Segment/SegmentAsync.tsx:
```diff
@@ -28,12 +30,25 @@ export function SegmentAsync<T>({
         Component={
           Component || (
-            <a className={cx('gf-form-label', 'query-part', !value && placeholder && 'query-placeholder', className)}>
+            <a
```
Using `a` without href is a bad accessibility practice, plus it seems like we're not using it as a link here at all, so can just change it to `span` and apply existing styles.

oscarkilhed committed:
Change from anchor to span in Segment pickers

Also value cannot be compared to undifined as empty string is a legit initial value for the picker

oscarkilhed committed:
Change clickable element in all segment pickers to spans

Also adds styling to make the spans indicate that they are clickable

Clarity-89 reviewed (approved):
Looks good! 🎉  Left a few suggestions. Also I think it'd be a good idea to enable the controls for the components story in the storybook so it's easier to test props visually.

oscarkilhed commented on packages/grafana-ui/src/components/Segment/styles.ts:
```diff
@@ -0,0 +1,15 @@
+import { css } from 'emotion';
+import { stylesFactory } from '../../themes';
+
+export const getStyles = stylesFactory(() => {
```
Absolutely!

oscarkilhed commented on packages/grafana-ui/src/components/Segment/Segment.tsx:
```diff
@@ -18,19 +19,34 @@ export function Segment<T>({
         Component={
           Component || (
-            <a className={cx('gf-form-label', 'query-part', !value && placeholder && 'query-placeholder', className)}>
+            <span
+              className={cx(
+                'gf-form-label',
+                'query-part',
+                styles.clickable,
+                {
+                  ['query-placeholder']: placeholder !== undefined && !value,
+                  [styles.disabled]: disabled,
+                },
+                className
+              )}
+            >
```
Sounds like a good idea.

oscarkilhed committed:
Change to InlineLabel from span

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

oscarkilhed committed:
Use InlineLabel instead of span and remove some sass-classes in Segments

oscarkilhed committed:
Change MetricsQueryEditor test to reflect new layout of AsyncSegment

Clarity-89 commented on packages/grafana-ui/src/components/Segment/Segment.tsx:
```diff
@@ -18,19 +21,33 @@ export function Segment<T>({
 
   if (!expanded) {
     const label = _.isObject(value) ? value.label : value;
+    const theme = useTheme();
+    const styles = getSegmentStyles(theme);
```
```suggestion
    const styles = useStyles(getSegmentStyles);
```
If get styles function has only theme param, we can use `useStyles` hook: https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/themes/ThemeContext.tsx#L51, which would provide the theme and memoize `getStyles` calls.

oscarkilhed replied:
Ah, thanks!

Clarity-89 reviewed (approved):
Good job! One minor suggestion and I think this is good to go.

oscarkilhed committed:
Use useStyles hook to get themed styles for segment inputs

torkelo reviewed (approved):
Merge it!
